README
Prima-Image-Magick version 0.01
==============================

Prima::Image::Magick - Juggle images between Prima and Image::Magick

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

  Prima
  Image::Magick

COPYRIGHT AND LICENCE

Copyright (C) 2007 by Dmitry Karasik

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Prima::Image::Magick.

0.01  - original version

0.02  - copy Image::Magick namespace into Prima::Image

0.03  - build on win32

0.04-5  - upgrade for recent versions of ImageMagick

0.06  - force-convert 16-bit (and higher) image data into 8-bits


mag.c
/* $Id: mag.c,v 1.1.1.1 2007/03/16 17:16:14 dk Exp $ */

#ifdef __cplusplus
extern "C" {
#endif

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include <magick/MagickCore.h>
#include "mag.h"

void magick_croak( char * what, ExceptionInfo * exception )
{
        char message[MaxTextExtent] = "unknown exception";
        if ((exception)->severity != UndefinedException)
                FormatMagickString( 
                        message, 
                        MaxTextExtent,
                        "Exception %d: %s%s%s%s",
                        (exception)->severity, 
                        (exception)->reason ?
                                GetLocaleExceptionMessage(
                                        (exception)->severity,
                                        (exception)->reason
                        ) : "Unknown", 
                        (exception)->description ? " (" : "",
                        (exception)->description ? GetLocaleExceptionMessage(
                                (exception)->severity,
                                (exception)->description
                        ) : "",
                        (exception)->description ? ")" : ""
                );
	DestroyExceptionInfo( exception);
        croak("%s: %s", what, message);
}


#ifdef __cplusplus
}
#endif


prima.c
/* $Id: prima.c,v 1.4 2012/01/03 16:45:33 dk Exp $ */
#ifdef __cplusplus
extern "C" {
#endif

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include "prima.h"
#include <apricot.h>
#include <Image.h>
#include <img_conv.h>

void
read_prima_image_data( SV * input, pim_image * pim)
{
	int want_resample = 0;
	PImage i = ( PImage) gimme_the_mate( input);
	if ( i == NULL) croak("Cannot read Prima::Image data");

	pim-> data      = i-> data;
	pim-> line_size = i-> lineSize;
	pim-> width     = i-> w;
	pim-> height    = i-> h;
	pim-> bpp       = i-> type & imBPP;
	pim-> colors    = i-> palSize;
	pim-> palette   = (unsigned char*) i-> palette;
	pim-> category  = 0; 
	if ( i-> type & imGrayScale)
		pim-> category  |= IS_GRAY;
		want_resample = 1;
	if ( i-> type & imComplexNumber) {
		pim-> category  |= ( pim-> bpp == 16 * sizeof( float)) ? IS_FLOAT : IS_DOUBLE;
		pim-> category  |= IS_COMPLEX;
		want_resample = 1;
	} else if ( i-> type & imRealNumber) {
		pim-> category  |= ( pim-> bpp == 8 * sizeof( float)) ? IS_FLOAT : IS_DOUBLE;
		want_resample = 1;
	}
		
	if ( want_resample) {
		double srcLo = i-> self-> stats(( Handle) i, 0, isRangeLo, 0);
		double srcHi = i-> self-> stats(( Handle) i, 0, isRangeHi, 0);
		double dstHi = 255;
		double dstLo = 0;
	
		if ( srcHi != srcLo ) {
			pim-> resample_coeff_a = (dstHi - dstLo) / ( srcHi - srcLo );
			pim-> resample_coeff_b = (dstLo * srcHi - dstHi * srcLo) / ( srcHi - srcLo );
		} else {
			pim-> resample_coeff_a = 1.0;
			pim-> resample_coeff_b = 0.0;
		}
	}
}


void bytecopy  (pim_image*pim,void *s,void *d,int w) { memcpy          ( d, s, w); }
void bitexp    (pim_image*pim,void *s,void *d,int w) { bc_mono_graybyte( s, d, w, (PRGBColor)(pim->palette)); }
void nibbleexp (pim_image*pim,void *s,void *d,int w) { bc_nibble_graybyte( s, d, w, (PRGBColor)(pim->palette)); }
void bitrgb    (pim_image*pim,void *s,void *d,int w) { bc_mono_rgb     ( s, d, w, (PRGBColor)(pim->palette)); }
void nibblergb (pim_image*pim,void *s,void *d,int w) { bc_nibble_rgb   ( s, d, w, (PRGBColor)(pim->palette)); }
void bytergb   (pim_image*pim,void *s,void *d,int w) { bc_byte_rgb     ( s, d, w, (PRGBColor)(pim->palette)); }

#define RESAMPLE { while (w--) *d++ = *s++ * pim-> resample_coeff_a + pim-> resample_coeff_b + 0.5; }

void shorts(pim_image*pim,Short *s,Byte *d,int w)   RESAMPLE
void longs(pim_image*pim,Long *s,Byte *d,int w)   RESAMPLE
void floats(pim_image*pim,float *s,Byte *d,int w)   RESAMPLE
void doubles(pim_image*pim,double *s,Byte *d,int w) RESAMPLE

BitCopyProc *
get_prima_bitcopy_proc( int category, int bpp_from, int bpp_to )
{
	if ( category & IS_FLOAT) 
		return (BitCopyProc*) floats;
	if ( category & IS_DOUBLE) 
		return (BitCopyProc*) doubles;

	switch( bpp_to) {
	case 24:
		switch ( bpp_from) {
		case 1:
			return bitrgb;
		case 4:
			return nibblergb;
		case 8:
			return bytergb;
		}
		break;
	default:
		switch ( bpp_from) {
		case 1:
			return bitexp;
		case 4:
			return nibbleexp;
		case 16:
			return (BitCopyProc*) shorts;
		case 32:
			return (BitCopyProc*) longs;
		}
	}

	return bytecopy;
}

void
allocate_prima_image( SV * input, int width, int height, int rgb)
{
	PImage i = ( PImage) gimme_the_mate( input);
	if ( i == NULL) croak("Cannot read Prima::Image data");
	i-> self-> create_empty(( Handle) i, width, height, rgb ? imRGB : imByte);
}


#ifdef __cplusplus
}
#endif


prima.h
/* $Id: prima.h,v 1.2 2011/12/20 09:45:01 dk Exp $ */

#ifdef __cplusplus
extern "C" {
#endif

#define IS_GRAY     1
#define IS_FLOAT    2
#define IS_DOUBLE   4
#define IS_COMPLEX  8

typedef struct _pim_image {
	unsigned char * data;
	unsigned char * palette;
	unsigned int    colors;
	unsigned int    line_size;
	unsigned int    width;
	unsigned int    height;
	unsigned int    bpp;
	unsigned int    category;
	
	double   resample_coeff_a;
	double   resample_coeff_b;
} pim_image;

typedef void BitCopyProc( pim_image * pim, void * src, void * dst, int width);

void
read_prima_image_data( SV * input, pim_image * pim);

BitCopyProc *
get_prima_bitcopy_proc( int category, int bpp_from, int bpp_to );

void
allocate_prima_image( SV * input, int width, int height, int rgb);

#ifdef __cplusplus
}
#endif


Makefile.PL
# $Id: Makefile.PL,v 1.5 2012/01/03 16:45:33 dk Exp $
use 5.008008;
use ExtUtils::MakeMaker;

my ( $i, $d, $l) = ('','','');
eval { 
	require Prima::Config;
	$d = join(' ', map { "-D$_" } @{$Prima::Config::Config{cdefs}});
	$i = join(' ', map { "-I$_" } @{$Prima::Config::Config{incpaths}});
        for ( '/usr/include', @{$Prima::Config::Config{incpaths}}) {
                $i .= " -I$_/ImageMagick" if -d "$_/ImageMagick";
        }
	$l = $Prima::Config::Config{libname} if $^O =~ /win32|os2/i;
};

my $libname;
if ( $^O =~ /win32/i) {
	$libname = 'CORE_RL_magick_';
} else {
	$libname = 'MagickCore';
	if ( 
		`convert -version` =~ /Version: ImageMagick (\d+\.\d+\.\d+)/ 
	) {
		my $v = $1;
		$v =~ s/\.//g;
		$libname = 'Magick' if $v < 641;
	}
}

WriteMakefile(
	NAME              => 'Prima::Image::Magick',
	VERSION_FROM      => 'lib/Prima/Image/Magick.pm',
	PREREQ_PM         => {
		'Prima'		=> 1,
		'Image::Magick' => 1,
	},
	ABSTRACT_FROM     => 'lib/Prima/Image/Magick.pm',
	AUTHOR            => 'Dmitry Karasik <dmitry@karasik.eu.org>',
	LIBS              => [ "$l -l$libname" ],
	DEFINE            => $d,
	INC               => $i,
	OBJECT            => "prima\$(OBJ_EXT) mag\$(OBJ_EXT) Magick\$(OBJ_EXT)",
);


example/example.pl
use lib qw(./blib/lib ./blib/arch);
use Prima qw(Application ImageViewer Label);
use Image::Magick;
use Prima::Image::Magick qw(:all);

my $p = Prima::Image-> new( width => 10, height => 10);

$p-> begin_paint;
$p-> clear;
$p-> line(0,0,9,9);
$p-> line(0,9,9,0);
$p-> end_paint;

my @types;
my $o = $p-> dup;
$o-> size( 100, 100);
push @types, [ $o, 'linear' ];

my $adaptive = $p-> dup;
$adaptive-> AdaptiveResize( width => 100, height => 100);
push @types, [ $adaptive, 'adaptive' ];

my $gaussian = $p-> dup;
$gaussian-> Resize( width => 100, height => 100, filter => 'Gaussian');
push @types, [ $gaussian, 'gaussian' ];

my $cubic = $p-> dup;
$cubic-> Resize( width => 100, height => 100, filter => 'Cubic');
push @types, [ $cubic, 'cubic' ];

Prima::MainWindow-> new(
	text => 'Prima::Image::Magick demo',
)-> insert( map { 
	[ 'Prima::Label'       => 
		pack  => { expand => 1, fill => 'both' }, 
		text  => "$$_[1] scaling"
	], [ 'Prima::ImageViewer' => 
		pack  => { expand => 1, fill => 'both' }, 
		image => $$_[0]
	] } @types
);

run Prima;


lib/Prima/Image/Magick.pm
# $Id: Magick.pm,v 1.6 2012/01/03 16:47:16 dk Exp $
package Prima::Image::Magick;

use strict;
use warnings;
require Exporter;
our @ISA         = qw(Exporter);
our @EXPORT_OK   = qw( prima_to_magick magick_to_prima );
our %EXPORT_TAGS = ( all => \@EXPORT_OK );
our $VERSION     = '0.06';

require XSLoader;
XSLoader::load('Prima::Image::Magick', $VERSION);

use Prima;
use Image::Magick;

# proxy Image::Magick methods into Prima::Image
{
	my $v = join('|', @Image::Magick::EXPORT);
	my %d = map { $_ => 1 } grep { !/^([a-z_].*|[A-Z_]+|$v|Prima)$/ } keys %Image::Magick::;
	# delete aliases
	for my $meth ( keys %d) {
		if ( exists $d{"${meth}Image"}) {
			delete $d{"${meth}Image"};
			next;
		}
	}
	for my $meth ( keys %d) {
		next if exists $Prima::Image::{$meth};
		my $sub = sub {
			my $self = shift;
			my $mag  = $self-> BeginMagick;
			my @res;
			if ( wantarray) {
				@res = $mag-> $meth( @_ );
			} else {
				$res[0] = $mag-> $meth( @_ );
			}
			$self-> EndMagick;
			return wantarray ? @res : $res[0];
		};
		no strict 'refs';
		*{"Prima::Image::$meth"} = $sub;
	}
}

sub Prima::Image::BeginMagick
{
	my $self = $_[0];
	if ( exists $self-> {__ImageMagickStorage} ) {
		$self-> {__ImageMagickStorage}-> [0]++;
	} else {
		$self-> {__ImageMagickStorage} = [ 0, prima_to_magick( $self ) ];
	}
	$self-> {__ImageMagickStorage}-> [1];
}

sub Prima::Image::EndMagick
{
	my $self = $_[0];
	return unless 
		exists $self-> {__ImageMagickStorage} and 
		0 == $self-> {__ImageMagickStorage}->[0]--;
	convert_to_prima( $self-> {__ImageMagickStorage}->[1], $self);
	delete $self-> {__ImageMagickStorage};
}

sub prima_to_magick
{
	my ( $p) = @_;
	die "Not a Prima::Image object" unless $p and $p->isa('Prima::Image');
	my $m = Image::Magick-> new();
	convert_to_magick( $p, $m);
	$m;
}

sub magick_to_prima
{
	my ( $m, %h) = @_;
	die "Not an Image::Magick object" unless $m and $m->isa('Image::Magick');
	my $p = Prima::Image-> new;
	convert_to_prima( $m, $p);
	$p;
}

*Prima::Image::Magick = \&prima_to_magick;
*Image::Magick::Prima = \&magick_to_prima;

1;
__END__

=head1 NAME

Prima::Image::Magick - Juggle images between Prima and Image::Magick

=head1 SYNOPSIS

  use Prima::Image::Magick;

  my $i = Prima::Image-> new( ... );  # native prima images
  $i-> MedianFilter( radius => 5);    # can call Image::Magick methods

=head1 DESCRIPTION

Allows transformations between L<Prima> images and L<Image::Magick> images.
Exports all methods found on C<Image::Magick> into C<Prima::Image> space, thus
opening the possibilities of ImageMagick for Prima images.

=head1 Prima::Image API

The mutator methods found on C<Image::Magick> namespace are wrapped and
imported into C<Prima::Image> space, so that an image is implictly converted
to C<Image::Magick> and back, so that for example

    $prima_image-> Edge( radius => 5);

is actually the same as

    my $m = prima_to_magick( $prima_image);
    $m-> Edge( radius => 5);
    $prima_image = magick_to_prima( $m);

except that C<$prima_image> internally remains the same perl object. 

This approach is obviusly ineffective when more than one call to ImageMagick
code is required. To avoid the ineffeciency, wrappers C<BeginMagick> and C<EndMagick>
are declared, so 

    $prima_image-> BeginMagick;
    $prima_image-> Edge( radius => 5);
    $prima_image-> Enhance;
    $prima_image-> EndMagick;

is same as
    
    my $m = prima_to_magick( $prima_image);
    $m-> Edge( radius => 5);
    $m-> Enhance;
    $prima_image = magick_to_prima( $m);


=head1 Prima::Image::Magick API

=over

=item prima_to_magick $magick_image

Returns a deep copy of C<$magick_image> stored in a new instance of C<Prima::Image>
object. C<$magick_image> must contain exactly one ImageMagick bitmap. This means that
empty objects and objects f.ex. after C< Read('file1', 'file2') > cannot be used here.
Use C<Image::Magick::Deconstruct> to break image sequence into constituent parts.

Exported either by explicit request or as a part of C<use Prima::Image::Magick ':all'> call.

=item prima_to_magick $prima_image

Returns a deep copy of C<$prima_image> stored in a new instance of C<Image::Magick>
object.

Exported either by explicit request or as a part of C<use Prima::Image::Magick ':all'> call.

=item convert_to_magick $prima_image, $magick_image

Copies content of C<$prima_image> to C<$magick_image>. Not to be called directy
unless really necessary; the method is not exported, and its syntax may change
in future. 

=item convert_to_prima $magick_image, $prima_image

Copies content of C<$magick_image> to C<$prima_image>. Not to be called directy
unless really necessary; the method is not exported, and its syntax may change
in future.

=back

=head1 SEE ALSO

L<Prima>, L<Image::Magick>, L<examples/example.pl>.

=head1 AUTHOR

Dmitry Karasik, E<lt>dmitry@karasik.eu.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Dmitry Karasik

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
